Add psalm annotation, move to iterable, fix deprecation

fezfez opened the pull request:
- Add psalm annotation :sparkles: 
- move to iterable instead of Traversable :nerd_face: 
- fix deprecation from doctrine/orm 2.8 :1st_place_medal: 
- psalm totally typed = true 

Note : Im not 100% that psalm annotation are correctly implemented

fixes #89
fixes #71

fezfez committed:
Add psalm annotation, move to iterable, fix deprecation

fezfez commented:
hello @Ocramius and happy new year !
Do you have time to review this ?

Thanks !

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:
```diff
@@ -20,32 +19,49 @@
     public static function fromQuery(AbstractQuery $query, int $batchSize): self
     {
-        return new self($query->iterate(), $query->getEntityManager(), $batchSize);
+        return new self($query->toIterable(), $query->getEntityManager(), $batchSize);
```
I think this requires raising minimum ORM version dependency

fezfez replied:
doctrine/orm is already ^2.8. see : https://github.com/Ocramius/DoctrineBatchUtils/blob/8affefde661133a385c2416dcb63277e77e1913b/composer.json#L22

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:
```diff
@@ -20,32 +19,49 @@
 
     /**
-     * @param Traversable<mixed> $results
+     * @param Traversable<E, F> $results
```
Given this template, shouldn't the return type be `self<E, F>`?

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:
```diff
@@ -20,32 +19,49 @@
 
     /**
-     * @param mixed[] $results
+     * @param array<C, D> $results
```
Return type should be `self<C, D>`

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:
```diff
@@ -20,32 +19,49 @@
  * every batch size to 'detach' the managed objects and therefore make them entities only for the
  * purpose of reading without manually attaching them back to the EntityManager.
+ *
+ * @template A
```
Let's call these TKey, TValue rather than A and B

Ocramius replied:
I think `A` should be constrained to `int|string`

fezfez replied:
If i constraint TKey to int|string i got this error

``` 
ERROR: MixedArgumentTypeCoercion - src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:39:25 - Argument 1 of DoctrineBatchUtils\BatchProcessing\SelectBatchIteratorAggregate::__construct expects iterable<int|string, mixed>, parent type iterable<mixed, mixed> provided (see https://psalm.dev/194)
        return new self($query->toIterable(), $query->getEntityManager(), $batchSize)
```

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php:
```diff
@@ -56,7 +72,7 @@ public static function fromTraversableResult(
 
     /**
-     * {@inheritDoc}
+     * @return Traversable<A, B|mixed>
```
Wait... Why `|mixed`? 🤔

fezfez replied:
same as https://github.com/Ocramius/DoctrineBatchUtils/pull/99#discussion_r551008629

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SimpleBatchIteratorAggregate.php:
```diff
@@ -17,28 +16,46 @@
 use function key;
 
+/**
+ * @template A
```
TKey, TValue instead of A, B

Ocramius replied:
I think `A` should be constrained to `int|string`

fezfez replied:
same as https://github.com/Ocramius/DoctrineBatchUtils/pull/99#discussion_r551004131

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SimpleBatchIteratorAggregate.php:
```diff
@@ -17,28 +16,46 @@
 
     /**
-     * @param mixed[] $results
+     * @param array<C, D> $results
```
Return type to be defined here

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SimpleBatchIteratorAggregate.php:
```diff
@@ -17,28 +16,46 @@
 
     /**
-     * @param Traversable<mixed> $results
+     * @param Traversable<E, F> $results
```
Return type to be defined here

Ocramius commented on src/DoctrineBatchUtils/BatchProcessing/SimpleBatchIteratorAggregate.php:
```diff
@@ -49,7 +66,7 @@ public static function fromTraversableResult(
 
     /**
-     * {@inheritDoc}
+     * @return Traversable<A, B|mixed>
```
Why `|mixed`? That defeats the whole typing 🤔

fezfez replied:
i think i made this because of this error.

``` 
INFO: InvalidReturnType - src/DoctrineBatchUtils/BatchProcessing/SimpleBatchIteratorAggregate.php:71:16 - The declared return type 'Traversable<TKey:DoctrineBatchUtils\BatchProcessing\SimpleBatchIteratorAggregate as mixed, TValue:DoctrineBatchUtils\BatchProcessing\SimpleBatchIteratorAggregate as mixed>' for DoctrineBatchUtils\BatchProcessing\SimpleBatchIteratorAggregate::getIterator is incorrect, got 'Generator<TKey:DoctrineBatchUtils\BatchProcessing\SimpleBatchIteratorAggregate as mixed, (TValue:DoctrineBatchUtils\BatchProcessing\SimpleBatchIteratorAggregate as mixed)|object, mixed, void>' (see https://psalm.dev/011)
     * @return Traversable<TKey, TValue>
``` 

and psalm propose to fix as 

```diff
@@ -68,9 +68,11 @@
     }
 
     /**
-     * @return Traversable<TKey, TValue>
+     * @return \Generator
+     *
+     * @psalm-return \Generator<TKey, TValue|object, mixed, void>
      */
-    public function getIterator(): iterable
+    public function getIterator(): \Generator
     {
         $iteration = 0;
         $resultSet = $this->resultSet;

``` 

I think the part `TValue|object, mixed, void` is because of  how `getIterator` is design (re fetch object, if its array take the first elements etc...).
I am not a an pslam expert so have you any idea ?

fezfez replied:
Ping!

Ocramius reviewed (changes requested):
Sorry, missed this! Thanks for the ping 👍

fezfez committed:
Fix after review

fezfez committed:
increase phpstan level

fezfez committed:
add @implements

Ocramius commented:
Yeah, I didn't have time to inspect this yet.

This requires patient inspection of your patch on my local machine.

On Sun, Jan 3, 2021, 13:32 Stéphane <notifications@github.com> wrote:

> *@fezfez* commented on this pull request.
> ------------------------------
>
> In src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php
> <https://github.com/Ocramius/DoctrineBatchUtils/pull/99#discussion_r550999942>
> :
>
> >      public static function fromQuery(AbstractQuery $query, int $batchSize): self
>      {
> -        return new self($query->iterate(), $query->getEntityManager(), $batchSize);
> +        return new self($query->toIterable(), $query->getEntityManager(), $batchSize);
>
> doctrine/orm is already ^2.8. see :
> https://github.com/Ocramius/DoctrineBatchUtils/blob/8affefde661133a385c2416dcb63277e77e1913b/composer.json#L22
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Ocramius/DoctrineBatchUtils/pull/99#discussion_r550999942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFVEHFOBI6JMQASHHMDMTSYBPV3ANCNFSM4U4D7RBQ>
> .
>

Ocramius committed:
Dropped `phpstan/*` dependencies - for now, not worth the effort to investigate double CI

Sadly, this is a side-effect of slow dependency upgrades and missing dependabot.

Hopefully, we'll re-introduce it later, when available
in https://github.com/laminas/laminas-ci-matrix-action

Ocramius committed:
Removed redundant `@var class-string`, now correctly infered from `doctrine/persistence`

Ocramius committed:
Stricter type checks for `reFetchObject()` internals

Removed a `@psalm-assert`, replaced with `@param` and internal type documentation
about the quirks of DQL iteration with aliases.

Ocramius committed:
Applied automated CS fixes

Ocramius reviewed (approved):
@fezfez sorry, I only got to this library now.

I did a full cleanup of the branch, and simplified some details, so it's ready for shipping/releasing.

I'll add PHP 8.x before releasing though.

Sorry it took so long :-(
